From 1e1e66f65675343567b47406ceb4d4ff50a25dce Mon Sep 17 00:00:00 2001
From: Dan Horak <dan@danny.cz>
Date: Sun, 20 Jul 2008 09:14:30 +0200
Subject: [PATCH] s390-tools-1.5.3-lvm

---
 fdasd/fdasd.c |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/fdasd/fdasd.c b/fdasd/fdasd.c
index d011b69..9101c01 100644
--- a/fdasd/fdasd.c
+++ b/fdasd/fdasd.c
@@ -280,6 +280,8 @@ fdasd_partition_type (char *str)
 		strcpy(str, "Linux swap");
 	else if (strncmp("RAID  ", str, 6) == 0)
 		strcpy(str, "Linux Raid");
+	else if (strncmp("LVM   ", str, 6) == 0)
+		strcpy(str, "Linux LVM");
 	else
 		strcpy(str, "unknown");
 
-- 
1.6.0.6

