From 77f053260b9d2b4d683edfbed50a528d74620d4b Mon Sep 17 00:00:00 2001
From: =?utf-8?q?Dan=20Hor=C3=A1k?= <dan@danny.cz>
Date: Fri, 24 Apr 2009 14:05:29 +0200
Subject: [PATCH] remove the executable stack flag

---
 zipl/src/Makefile |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/zipl/src/Makefile b/zipl/src/Makefile
index 16d5276..07b3b74 100644
--- a/zipl/src/Makefile
+++ b/zipl/src/Makefile
@@ -12,7 +12,7 @@ includes = $(wildcard ../include/*.h)
 all: zipl
 
 zipl: $(objects)
-	$(CC) $(objects) ../boot/data.o -o zipl
+	$(CC) -Wl,-z,noexecstack $(objects) ../boot/data.o -o zipl
 
 %.o: %.c $(includes) Makefile
 	$(CC) $(CFLAGS) -c -o $@ $<
-- 
1.6.0.6

