From c4a38de57376a6ddf03906afeac142525837aab0 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Dan=20Hor=C3=A1k?= <dan@danny.cz>
Date: Fri, 18 Mar 2011 16:35:17 +0100
Subject: [PATCH 64/66] switch to using udevadm settle

---
 etc/init.d/mon_statd |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/etc/init.d/mon_statd b/etc/init.d/mon_statd
index 60bcf00..b6699c7 100755
--- a/etc/init.d/mon_statd
+++ b/etc/init.d/mon_statd
@@ -39,7 +39,7 @@ load_kernel_module()
 		if [ $? -ne 0 ]; then
 			exit 1
 		fi
-		udevsettle
+		udevadm settle
 		if [ $? -ne 0 ]; then
 			exit 1
 		fi
-- 
1.7.4

