diff -up s390-tools-1.8.2/ziomon/Makefile.linker s390-tools-1.8.2/ziomon/Makefile
--- s390-tools-1.8.2/ziomon/Makefile.linker	2010-06-10 08:54:55.000000000 +0200
+++ s390-tools-1.8.2/ziomon/Makefile	2010-06-10 08:55:48.000000000 +0200
@@ -21,7 +21,7 @@ ziomon_util: ziomon_util_main.o ziomon_t
 ziomon_zfcpdd_main.o: ziomon_zfcpdd.c ziomon_zfcpdd.h
 	$(CC) -DWITH_MAIN $(CFLAGS) $(CPPFLAGS) -c $< -o $@
 ziomon_zfcpdd: ziomon_zfcpdd_main.o ziomon_tools.o
-	$(LINK) $^ -o $@ -lm -lrt
+	$(LINK) $^ -o $@ -lm -lrt -lpthread
 
 ziorep_traffic: ziorep_traffic.o ziorep_framer.o ziorep_frameset.o \
 		ziorep_printers.o ziomon_dacc.o ziomon_util.o \
