From 804e5362c0205f4fd1c060cae9b6ac5972d097a7 Mon Sep 17 00:00:00 2001
From: Dan Horak <dan@danny.cz>
Date: Sun, 20 Jul 2008 09:24:05 +0200
Subject: [PATCH 1/4] s390-tools-1.5.3-zipl-zfcpdump-2

---
 common.mak |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/common.mak b/common.mak
index 589deb4..8ab0225 100644
--- a/common.mak
+++ b/common.mak
@@ -62,8 +62,8 @@ GROUP		= $(shell id -gn)
 export INSTROOT BINDIR LIBDIR MANDIR OWNER GROUP
 
 # Special defines for zfcpdump
-ZFCPDUMP_DIR    = /usr/local/share/zfcpdump
-ZFCPDUMP_IMAGE  = zfcpdump.image
+ZFCPDUMP_DIR    = /boot
+ZFCPDUMP_IMAGE  = zfcpdump
 ZFCPDUMP_RD     = zfcpdump.rd
 export ZFCPDUMP_DIR ZFCPDUMP_IMAGE ZFCPDUMP_RD
 
-- 
1.7.4.4


From dfde1173c937c1bf7a344bcc21ac51fb9b9ca6c1 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Dan=20Hor=C3=A1k?= <dan@danny.cz>
Date: Thu, 23 Apr 2009 11:46:01 +0200
Subject: [PATCH 2/4] s390-tools-1.8.1-fdasd-su

---
 fdasd/fdasd.c |   10 ++++++----
 1 files changed, 6 insertions(+), 4 deletions(-)

diff --git a/fdasd/fdasd.c b/fdasd/fdasd.c
index d8cf584..7a61022 100644
--- a/fdasd/fdasd.c
+++ b/fdasd/fdasd.c
@@ -2093,10 +2093,12 @@ fdasd_get_geometry (fdasd_anchor_t *anc)
 	if (anc->verbose) printf("disk type check     : ok\n");
 
         if (dasd_info.FBA_layout != 0) {
-                snprintf(err_str, ERROR_STRING_SIZE, 
-			"%s is not formatted with z/OS compatible "
-			"disk layout!", options.device);
-                fdasd_error(anc, wrong_disk_format, err_str);
+                if(!anc->silent) {
+                    snprintf(err_str, ERROR_STRING_SIZE,
+			    "%s is not formatted with z/OS compatible "
+			    "disk layout!", options.device);
+                    fdasd_error(anc, wrong_disk_format, err_str);
+                }
         }      
 
 	if (anc->verbose) printf("disk layout check   : ok\n");
-- 
1.7.4.4


From f8a90e35a49848bfdcc4c984567b83fcb5df379a Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Dan=20Hor=C3=A1k?= <dan@danny.cz>
Date: Fri, 19 Jun 2009 10:01:30 +0200
Subject: [PATCH 3/4] s390-tools-1.8.1-zipl-kdump-man

Description: Add kdump kernel installation instruction to zipl man page.
Symptom:     User wants to prepare SCSI disk for dump, but has not installed
             the kdump kernel rpm.
Problem:     The installation of the kdump kernel rpm is prereq for preparing
             a SCSI dump disk for dump.
Solution:    Document that in the zipl man page.
---
 zipl/man/zipl.8 |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/zipl/man/zipl.8 b/zipl/man/zipl.8
index 4c12d59..26b79cf 100644
--- a/zipl/man/zipl.8
+++ b/zipl/man/zipl.8
@@ -263,6 +263,8 @@ will be incomplete.
 It is not possible to specify both this parameter and the name of a menu
 or configuration section on the command line at the same time.
 
+.B Note that before using this option the "kernel-kdump" rpm has to be
+.B installed.
 .TP
 .BR "\-M <DUMPLIST[,SIZE]>" " or " "--mvdump=<DUMPLIST[,SIZE]>"
 Install a multi-volume dump record on each device associated with one of the
-- 
1.7.4.4


From 30e79c2abe06873666b7fc159eb2643bad61ef96 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Dan=20Hor=C3=A1k?= <dan@danny.cz>
Date: Fri, 18 Mar 2011 16:35:17 +0100
Subject: [PATCH 4/4] switch to using udevadm settle

---
 etc/init.d/mon_statd |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/etc/init.d/mon_statd b/etc/init.d/mon_statd
index 60bcf00..b6699c7 100755
--- a/etc/init.d/mon_statd
+++ b/etc/init.d/mon_statd
@@ -39,7 +39,7 @@ load_kernel_module()
 		if [ $? -ne 0 ]; then
 			exit 1
 		fi
-		udevsettle
+		udevadm settle
 		if [ $? -ne 0 ]; then
 			exit 1
 		fi
-- 
1.7.4.4

